viewer/packages/cad-parsers/src/utilities/SectorScene.ts

```diff
@@ -88,60 +88,114 @@ export class SectorSceneImpl implements SectorScene {
     return accepted;
   }
 
+  private detectIntersectingBounds(bounds: THREE.Box3, clusters: THREE.Box3[]): THREE.Box3[] {
+    const intersectingClusters: THREE.Box3[] = [];
+    const potentialJunkClusters: THREE.Box3[] = [];
+
+    clusters.forEach(cluster => {
+      if (bounds !== cluster) {
+        if (bounds.intersectsBox(cluster)) {
+          intersectingClusters.push(cluster);
+        } else {
+          potentialJunkClusters.push(cluster);
+        }
+      }
+    });
+
+    let outClusters = intersectingClusters;
+    if (potentialJunkClusters.length > 0) {
+      intersectingClusters.forEach(cluster => {
+        outClusters = outClusters.concat(this.detectIntersectingBounds(cluster, potentialJunkClusters));
+      });
+    }
+    return outClusters;
+  }
+
   private computeBoundsOfMostGeometry(): THREE.Box3 {
     if (this.root.children.length === 0) {
       return this.root.subtreeBoundingBox;
     }
 
-    // Determine all corners of the bboxes
+    // Find all leaf bounds and count overal rendercost
     const allBounds: THREE.Box3[] = [];
-    const corners: number[][] = [];
-    traverseDepthFirst(this.root, x => {
-      if (x.children.length === 0) {
-        corners.push(x.subtreeBoundingBox.min.toArray(), x.subtreeBoundingBox.max.toArray());
-        allBounds.push(x.subtreeBoundingBox, x.subtreeBoundingBox);
-      }
-      return true;
-    });
-    // Cluster the corners into four groups and determine bounds of each cluster
-    const numClusters = Math.min(corners.length, 4);
-    const clusters = skmeans(corners, numClusters, 'kmpp', 10);
-    const clusterCounts = new Array<number>(clusters.idxs.length).fill(0);
-    const clusterBounds = clusterCounts.map(_ => new THREE.Box3());
-    clusters.idxs.map(x => clusterCounts[x]++);
-    const biggestCluster = clusterCounts.reduce(
-      (max, count, idx) => {
-        if (count > max.count) {
-          max.count = count;
-          max.idx = idx;
+    {
```

Comment:
There are a lot of things happening in this function, and although I know it was pretty bad before, could we split it into smaller functions so it's easier to follow? :sweat_smile:

Reply:
Created generateClusters as this felt like something that would make sense to do in a separate function.